### PR TITLE
fix(expr): constant folder should run repeatly until stable

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Closes #issue

### CODE REVIEW

File: src/query/functions/tests/it/scalars/testdata/boolean.txt [hunk 1]
```diff
@@ -337,6 +337,27 @@ evaluation (internal):
 +--------+--------------------------------------------------------------------------+
 
 
+ast            : const_false AND CAST('1000' AS STRING) = 1000
+raw expr       : and(#0::Boolean, eq(CAST("1000" AS String), 1000_u16))
+checked expr   : and<Boolean, Boolean>(#0, eq<Int64, Int64>(CAST("1000" AS Int64), CAST(1000_u16 AS Int64)))
+optimized expr : false
```
Comment (andylokandy) [inline src/query/functions/tests/it/scalars/testdata/boolean.txt:343]: @sundy-li PTAL if this satisfies.

File: src/query/expression/src/evaluator.rs [hunk 2]
```diff
@@ -518,7 +518,23 @@ impl<'a, Index: ColumnIndex> ConstantFolder<'a, Index> {
         }
     }
 
+    /// Fold a single expression, returning the new expression and the domain of the new expression.
     pub fn fold(&self, expr: &Expr<Index>) -> (Expr<Index>, Option<Domain>) {
+        let mut old_expr = expr.clone();
+        loop {
+            let (new_expr, domain) = self.fold_once(&old_expr);
```
Comment (sundy-li) [inline src/query/expression/src/evaluator.rs:525]: maybe we should add a max count and log the error to avoid bugs of hang.

Comment (sundy-li) [reply src/query/expression/src/evaluator.rs:525]: we can use `1024 * 1024` which common expression depth will not be reached.

File: src/query/functions/tests/it/scalars/testdata/boolean.txt (hunk 1 above)
Comment (sundy-li) [reply src/query/functions/tests/it/scalars/testdata/boolean.txt:343]: Great!

**Review by sundy-li — APPROVED** [review]

**Pull request merged**